docs/reference/numbers.tex

```diff
@@ -0,0 +1,601 @@
+template<>
+struct is_number<std::chrono::year> : std::true_type {};
+template<>
+struct is_number<std::chrono::weekday> : std::true_type {};
+template<>
+struct is_number<std::chrono::year_month> : std::true_type {};
+\end{itemdecl}
+
+\begin{itemdecl}
+template<typename T>
+struct @\libglobal{is_complex_number}@ : std::false_type {};
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\cvalue
+\tcode{true} if \tcode{T} represents a complex number\irefiev{102-02-09}, and
+\tcode{false} otherwise.
+\end{itemdescr}
+
+\indexlibraryspec{is_complex_number}{std::complex}%
+\begin{itemdecl}
+template<class T>
+struct is_complex_number<std::complex<T>> : std::true_type {};
+\end{itemdecl}
+
+\Sec2[num.ids]{Identities}
+
+\begin{codeblock}
+template<typename T>
+concept @\defexposconcept{inferable-identities}@ =
+  std::common_with<T, number_difference_t<T>> && std::constructible_from<T, int>;
+\end{codeblock}
+
+\indexlibrary{\idxcode{number_zero}!\idxexposconcept{inferable-identities}}%
+\begin{itemdecl}
+template<typename T>
+struct @\libglobal{number_zero}@ { @\seebelow@ };
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\cvalue
+\tcode{T}'s neutral element for addition\irefiev{102-01-12}, if any.
+
+\pnum
+\dvalue
+If \tcode{T} models \exposconcept{inferable-identities}, \tcode{T(0)}.
+\end{itemdescr}
+
+\indexlibrary{\idxcode{number_one}!\idxexposconcept{inferable-identities}}%
+\begin{itemdecl}
+template<typename T>
+struct @\libglobal{number_one}@ { @\seebelow@ };
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\cvalue
+\tcode{T}'s neutral element for multiplication\irefiev{102-01-19}, if any.
+
+\pnum
+\dvalue
+If \tcode{T} models \exposconcept{inferable-identities}, \tcode{T(1)}.
+\end{itemdescr}
+
+\Sec2[num.assoc.types]{Associated types}
+
+\begin{itemdecl}
+template<typename T>
+using @\libglobal{number_difference_t}@ = decltype(std::declval<const T>() - std::declval<const T>());
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\tcode{number_difference_t<T>} represents \tcode{T}'s difference's\irefiev{102-01-17} type, if any.
+\end{itemdescr}
+
+\indexlibrary{\idxcode{number_scalar}!arithmetic type}%
+\indexlibraryspec{number_scalar}{std::complex}%
+\indexlibraryspec{number_scalar}{std::chrono::duration}%
+\begin{itemdecl}
+template<typename T>
+struct @\libglobal{number_scalar}@ {};
+template<@\seebelow@>
+struct number_scalar<@\seebelow@> {
+  using type = @\seebelow@;
+};
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\ctype
+\tcode{T}'s scalar\irefiev{102-02-18}, if any.
+
+\pnum
+Each row of \tref{num.scalar} denotes a specialization.
+
+\begin{floattable}{Specializations of \tcode{number_scalar}}{num.scalar}{lll}
+\topline
+\lhdr{\fakegrammarterm{template-parameter-list}}
+                              & \chdr{\fakegrammarterm{template-argument-list}}
+                                                              & \rhdr{\Fundescx{Type}} \\ \rowsep
+\tcode{std::integral T}       & \tcode{T}                     & \tcode{T} \\ \rowsep
```

Comment:
I think it might semantically be a field number.
It works very well in practice thanks to integral promotion.

I have done things like
`assert(n.ids.begin() + 1 != n.ids.end() - int{exclude_last && n.ids.size() > 1})`
(using `int{}` for clarity).
As well as using `+` instead of `&&`,
i.e., `a + b` instead of `a && b`,
where that works better (for whatever reason I forgot).

Now, I just tried using `+=` and `-=`: <https://compiler-explorer.com/z/Wfb5j7sja>.
It seems like the `+=` has saturation semantics,
but `-=` wraps back to `true`.
This is all because only `0` is `false`,
all other values are `true`,
and `𝘰𝘱=` is defined as `x = x 𝘰𝘱 y;` (<https://eel.is/c++draft/expr.ass#6>).

Remember that these number concepts are performed in an unspecified set,
and the result is mapped back to the type of the result.
So the behavior of `bool`'s `+=` and `-=` satisfy the concepts.
(E.g., `false + 0` and `true - 1` map to `false`,
any other `int` input would map to `true`).

Now I understand that these concepts also support saturation arithmetic.